### PR TITLE
Fixes #27036 - improve logging of error backtraces

### PR DESCRIPTION
**Pull request by lzap**
Looks like it's not visible at all.

### CODE REVIEW

File: modules/dhcp_libvirt/subnet_service_initializer.rb [hunk 1]
```diff
@@ -37,7 +37,7 @@ def parse_config_for_subnets
       raise Proxy::DHCP::Error("Only one subnet is supported") if ret_val.size > 1
       ret_val
     rescue Exception => e
-      msg = "Unable to parse subnets XML"
+      msg = "Unable to parse subnets XML: #{e}"
       logger.error msg, e
```
Comment (tbrisker) [inline modules/dhcp_libvirt/subnet_service_initializer.rb]: Isn't the issue than actually that the logger.error ignores the second parameter and other places are also affected?

Comment (lzap) [reply modules/dhcp_libvirt/subnet_service_initializer.rb]: Yeah I was thinking the same this morning.

**Comment by lzap** [comment]
Ok improved it, actually regression made in: https://github.com/theforeman/smart-proxy/commit/8652a7e5c63009cae026f6350178583dcadc5f34

```
2019-06-14T08:56:28 9302404b [E] TEST
2019-06-14T08:56:28 9302404b [W] Error details for TEST: <StandardError>: Test
```

It's slightly more readable now.

After IRC discussion with @ekohl we agreed that adding support for `try` method from Rails will not hurt, it clearly makes code easier to read. After we upgrade to 2.3+ we can drop this.

**Comment by lzap** [comment]
Rebased, in case you can't boot smart-proxy with "daemon" dependency error - rebase. It was an issue we've fixed this morning.

File: lib/try-patch.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
+#
```
Comment (ekohl) [inline lib/try-patch.rb]: Given there's only one usage of this, I'd prefer to hold off on this. If you do have a larger block of code that uses it, then I'd be happy to include it.

**Comment by lzap** [comment]
Removed, rebased.

File: lib/proxy/log_buffer/decorator.rb [hunk 3]
```diff
@@ -38,6 +38,14 @@ def add(severity, message = nil, progname = nil, backtrace = nil)
           message = progname
         end
       end
+      # accepts backtrace, exception and simple string for historical reasons
+      backtrace = if exception_or_backtrace.is_a?(Exception) && !exception_or_backtrace.backtrace.nil?
```
Comment (ekohl) [inline lib/proxy/log_buffer/decorator.rb]: I don't see why you're moving this out here. It's only used on line 57 so nesting it inside the if there should be good enough?

Comment (lzap) [reply lib/proxy/log_buffer/decorator.rb]: Right, this was done for no particular reason. Rebased.

**Review by ekohl — COMMENTED** [review]
Could you add a test for `add`? This is complicated (for historical reasons) and I'd feel better if it had tests.

**Comment by lzap** [comment]
Rebased, added tests.

**Comment by lzap** [comment]
Fixed rubocop issue, ready for rereview.

File: modules/dhcp_libvirt/subnet_service_initializer.rb [hunk 4]
```diff
@@ -37,7 +37,7 @@ def parse_config_for_subnets
       raise Proxy::DHCP::Error("Only one subnet is supported") if ret_val.size > 1
       ret_val
     rescue Exception => e
-      msg = "Unable to parse subnets XML"
+      msg = "Unable to parse subnets XML: #{e}"
```
Comment (ekohl) [inline modules/dhcp_libvirt/subnet_service_initializer.rb:40]: Doesn't the `, e` to `logger.error` already add this information?

Comment (lzap) [reply modules/dhcp_libvirt/subnet_service_initializer.rb:40]: Yes but apparently I want the error message to be returned to the client. Our error reporting is sometimes very weak.

File: lib/proxy/log_buffer/decorator.rb [hunk 5]
```diff
@@ -46,15 +46,20 @@ def add(severity, message = nil, progname = nil, backtrace = nil)
         @logger.add(severity, message)
         # add add to the buffer
         if severity >= @logger.level
-          # we accept backtrace, exception and simple string
-          backtrace = backtrace.is_a?(Exception) ? backtrace.backtrace : backtrace
-          backtrace = backtrace.respond_to?(:join) ? backtrace.join("\n") : backtrace
+          # accepts backtrace, exception and simple string for historical reasons
+          backtrace = if exception_or_backtrace.is_a?(Exception) && !exception_or_backtrace.backtrace.nil?
+                        exception_or_backtrace.message + ': ' + exception_or_backtrace.backtrace.join("\n").to_s
```
Comment (tbrisker) [inline lib/proxy/log_buffer/decorator.rb]: not crucial but the `.to_s` here is redundent

**Review by tbrisker — APPROVED** [review]
Any further comments @ekohl? lgtm except for a super minor comment

**Comment by lzap** [comment]
Rebased.

File: modules/dhcp_libvirt/subnet_service_initializer.rb (hunk 4 above)
Comment (ekohl) [reply modules/dhcp_libvirt/subnet_service_initializer.rb:40]: Would `logger.exception` be useful here (and later in the file)?

**Review by ekohl — APPROVED** [review]
Would appreciate to know how the answer to my question but overall :+1: and can be merged regardless of the answer.

**Comment by lzap** [comment]
I can't comment in the file, so let me put it here. I can add that, this code is probably pre-exception helper. Rebased.

File: modules/dhcp_libvirt/subnet_service_initializer.rb (hunk 4 above)
Comment (ekohl) [reply modules/dhcp_libvirt/subnet_service_initializer.rb:40]: Is the specific exception message still needed when using `logger.exception`? I think that already formats the message to include the specific exception message.

**Comment by tbrisker** [comment]
please rebase @lzap

**Comment by lzap** [comment]
Rebased.

**Comment by tbrisker** [comment]
Thanks @lzap !

**Pull request merged**